cpp/src/parquet/chunker_internal_test.cc

```diff
@@ -0,0 +1,1397 @@
+      equal_diffs++;
+    } else if (original_sum < modified_sum) {
+      larger_diffs++;
+      ASSERT_EQ(original_sum + edit_length, modified_sum);
+    } else if (original_sum > modified_sum) {
+      smaller_diffs++;
+      ASSERT_EQ(original_sum, modified_sum + edit_length);
+    }
+    ASSERT_LE(diff.first.size(), 2);
+    ASSERT_LE(diff.second.size(), 2);
+  }
+
+  ASSERT_EQ(equal_diffs, exact_number_of_equal_diffs);
+  ASSERT_EQ(larger_diffs, exact_number_of_larger_diffs);
+  ASSERT_EQ(smaller_diffs, exact_number_of_smaller_diffs);
+}
+
+void AssertPageLengthDifferences(const RowGroupInfo& original,
+                                 const RowGroupInfo& modified,
+                                 uint8_t max_number_of_equal_diffs) {
+  // A less restrictive version of the above assertion function mainly used to
+  // assert the update case.
+  auto diffs = FindDifferences(original.page_lengths, modified.page_lengths);
+  if (diffs.size() > max_number_of_equal_diffs) {
+    PrintDifferences(original.page_lengths, modified.page_lengths, diffs);
+  }
+  ASSERT_LE(diffs.size(), max_number_of_equal_diffs);
+
+  for (const auto& diff : diffs) {
+    uint64_t left_sum = 0, right_sum = 0;
+    for (const auto& val : diff.first) left_sum += val;
+    for (const auto& val : diff.second) right_sum += val;
+    ASSERT_EQ(left_sum, right_sum);
+    ASSERT_LE(diff.first.size(), 2);
+    ASSERT_LE(diff.second.size(), 2);
+  }
+
+  if (diffs.size() == 0) {
+    // no differences found, the arrays are equal
+    ASSERT_TRUE(original.page_lengths == modified.page_lengths);
+  }
+}
+
+uint64_t ElementCount(int64_t size, int32_t byte_width, bool nullable) {
+  if (nullable) {
+    // in case of nullable types the def_levels are also fed through the chunker
+    // to identify changes in the null bitmap, this will increase the byte width
+    // and decrease the number of elements per chunk
+    byte_width += 2;
+  }
+  return size / byte_width;
+}
+
+void AssertAllBetween(const ChunkList& chunks, int64_t min, int64_t max) {
+  // except the last chunk since it is not guaranteed to be within the range
+  for (size_t i = 0; i < chunks.size() - 1; i++) {
+    ASSERT_GE(chunks[i], min);
+    ASSERT_LE(chunks[i], max);
+  }
+  ASSERT_LE(chunks.back(), max);
+}
+
+void AssertChunkSizes(const std::shared_ptr<::arrow::DataType>& dtype,
+                      const RowGroupInfo& base_info, const RowGroupInfo& modified_info,
+                      bool nullable, bool enable_dictionary, int64_t min_chunk_size,
+                      int64_t max_chunk_size) {
+  if (dtype->id() != ::arrow::Type::BOOL) {
+    ASSERT_EQ(base_info.has_dictionary_page, enable_dictionary);
+    ASSERT_EQ(modified_info.has_dictionary_page, enable_dictionary);
+  }
+  if (::arrow::is_fixed_width(dtype->id())) {
+    // for nullable types we cannot calculate the exact number of elements because
+    // not all elements are fed through the chunker (null elements are skipped)
```

Comment:
As said above, you can compute the exact number if you know the null count.

Reply:
Need to rewire the functions a little bit to pass the array instead of the dtype.

Reply:
Reworked it, now it properly calculates the cdc chunk size based on the page lengths and the original logical array. This is only applicable to the fixed width types, for list-like types I would need to reimplement the logic from the CDC chunker which wouldn't make a lot of sense.